### PR TITLE
Add ColumnEnsemble Forecaster

### PR DESCRIPTION
**Pull request by GuzalBulatova**
#### Reference Issues/PRs
References #1081. See also #923 and #854.

#### What does this implement/fix? Explain your changes.
Implements ColumnEnsebleForecaster.

#### Does your contribution introduce a new dependency? If yes, which one?
No.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

### CODE REVIEW

File: sktime/forecasting/compose/tests/test_column_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    ],
+)
+@pytest.mark.parametrize(
+    "fh", [(np.arange(1, 11)), (np.arange(1, 33)), (np.arange(1, 3))]
+)
+def test_column_ensemble_shape(forecasters, fh):
+    y = pd.DataFrame(np.random.randint(0, 100, size=(100, 3)), columns=list("ABC"))
+    forecaster = ColumnEnsembleForecaster(forecasters)
+    forecaster.fit(y)
+    fh = np.arange(1, 11)
```
Comment (mloning) [inline sktime/forecasting/compose/tests/test_column_ensemble.py:34]:
```suggestion

```

File: sktime/forecasting/compose/tests/test_column_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        ]
+    ],
+)
+@pytest.mark.parametrize(
+    "fh", [(np.arange(1, 11)), (np.arange(1, 33)), (np.arange(1, 3))]
+)
+def test_column_ensemble_shape(forecasters, fh):
+    y = pd.DataFrame(np.random.randint(0, 100, size=(100, 3)), columns=list("ABC"))
+    forecaster = ColumnEnsembleForecaster(forecasters)
+    forecaster.fit(y)
```
Comment (mloning) [inline sktime/forecasting/compose/tests/test_column_ensemble.py:33]:
```suggestion
    forecaster.fit(y, fh=fh)
```

File: sktime/forecasting/compose/tests/test_column_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    fh = np.arange(1, 11)
+    actual = forecaster.predict(fh=fh)
```
Comment (mloning) [inline sktime/forecasting/compose/tests/test_column_ensemble.py:35]:
```suggestion
    actual = forecaster.predict()
```

**Review by aiwalter — COMMENTED** [review]
left a few comments

File: sktime/forecasting/compose/_column_ensemble.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+from sktime.forecasting.base._base import BaseForecaster
+from sktime.forecasting.base._meta import _HeterogenousEnsembleForecaster
+
+
+class ColumnEnsembleForecaster(_HeterogenousEnsembleForecaster):
+    """Forecast each series with separate forecaster.
+
+    Parameters
+    ----------
+    forecasters : list of (str, estimator, index) tuples
```
[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py:23]: No, index can only be an int. I'm making these changes, thank you so much for feedback, Martin!

File: sktime/forecasting/compose/_column_ensemble.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    Example
+    -------
+    >>> from sktime.forecasting.compose import ColumnEnsembleForecaster
+    >>> from sktime.forecasting.exp_smoothing import ExponentialSmoothing
+    >>> from sktime.forecasting.trend import PolynomialTrendForecaster
+    >>> y = pd.DataFrame(np.random.randint(0, 100, size=(100, 2)))
```
[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py:30]: The reason I made this random one is that there's no example set as pd.DataFrame in sktime. Is there a better way to create an example for this case?

File: sktime/forecasting/compose/_column_ensemble.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    >>> forecasters = [("trend", PolynomialTrendForecaster(), 0),\
+                        ("ses", ExponentialSmoothing(trend='add'), 1)]
```
[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py:32]: I think this way it's easier to read. I looked into example notebooks and if many estimators are passed they look like this (e.g. 3.4.4 in 01_forecasting notebook).

File: sktime/forecasting/compose/_column_ensemble.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        """Set the parameters of this estimator.
+
+        Valid parameter keys can be listed with ``get_params()``.
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+        self._set_params("_forecasters", **kwargs)
+        return self
```
[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py:143]: Ah, great point! I've actually already moved it into _HeterogenousEnsembleForecaster and forgot to remove them from here, thanks! 👍

File: sktime/forecasting/compose/_column_ensemble.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aiwalter) [reply sktime/forecasting/compose/_column_ensemble.py:30]: if you need a `pd.DataFrame()` then do `_, X = load_longley()`

File: sktime/forecasting/compose/_column_ensemble.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aiwalter) [reply sktime/forecasting/compose/_column_ensemble.py:32]: yep its just a detail, but I think having `forecasters` and `forecaster` is not so easy to read also ;)

File: sktime/forecasting/compose/_column_ensemble.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    Parameters
+    ----------
+    forecasters : list of tuples (str, estimator, int)
+
+    Examples
+    --------
+    >>> from sktime.forecasting.compose import ColumnEnsembleForecaster
+    >>> from sktime.forecasting.exp_smoothing import ExponentialSmoothing
+    >>> from sktime.forecasting.trend import PolynomialTrendForecaster
+    >>> y = pd.DataFrame(np.random.randint(0, 100, size=(100, 2)))
```
Comment (aiwalter) [inline sktime/forecasting/compose/_column_ensemble.py:30]:
```suggestion
    >>> from sktime.datasets import load_longley
    >>> _, y = load_longley()
```

File: sktime/forecasting/compose/_column_ensemble.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/forecasting/compose/_column_ensemble.py]: yes, good spot - if the input is a data frame, it should be possible to refer to column names by a string.

[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py]: Got it @aiwalter and @fkiraly , I'm fixing it

[the remark this answers is not in the document]
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py]: UPD: will change it in another PR

File: sktime/forecasting/compose/_column_ensemble.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        -------
+        self : returns an instance of self.
+        """
+        names, forecasters, indices = self._check_forecasters()
+        if len(set(indices)) != len(indices):
+            raise ValueError(
+                "One estimator per column required. Found %s unique"
+                " estimators" % len(set(indices))
+            )
+        elif len(set(indices)) != len(y.columns):
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:96]: We need to make sure that the indices are complete for the passed data, so perhaps something like the following?
```python
np.array_equal(np.sort(indices), np.arange(len(y.columns)) 
```

File: sktime/forecasting/compose/_column_ensemble.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        return self
+
+    def _predict(self, fh=None, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
+
+        y_pred = np.zeros((len(fh), len(self.forecasters_)))
+        for (_, forecaster, index) in self.forecasters_:
+            y_pred[:, index] = forecaster.predict(fh)
+
+        y_pred = pd.DataFrame(data=y_pred)
+        y_pred.index = self.fh.to_absolute(self.cutoff)
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:117]: We should probably also make sure that we return the same column names as seen in `fit`.

File: sktime/forecasting/compose/_column_ensemble.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        return y_pred
+
+    def _check_forecasters(self):
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:120]: is this a re-implementation to handle the indices?

File: sktime/forecasting/compose/_column_ensemble.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        if (
+            self.forecasters is None
+            or len(self.forecasters) == 0
+            or not isinstance(self.forecasters, list)
+        ):
+            raise ValueError(
+                "Invalid 'estimators' attribute, 'estimators' should be a list"
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:127]:
```suggestion
                "Invalid 'forecasters' attribute, 'forecasters' should be a list"
```

File: sktime/forecasting/compose/_column_ensemble.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+                " of (string, estimator, int) tuples."
+            )
+        names, forecasters, indices = zip(*self.forecasters)
+        # defined by MetaEstimatorMixin
+        self._check_names(names)
+
+        has_estimator = any(est not in (None, "drop") for est in forecasters)
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:134]: I think this block of code can go

File: sktime/forecasting/compose/_column_ensemble.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        if not has_estimator:
+            raise ValueError(
+                "All estimators are dropped. At least one is required "
+                "to be an estimator."
+            )
+
+        for forecaster in forecasters:
+            if forecaster not in (None, "drop") and not isinstance(
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:142]:
```suggestion
            if not isinstance(
```

**Review by mloning — COMMENTED** [review]
Hi @GuzalBulatova - looks really good! I left a few comments. We also need to add it to the docs in `sktime/docs/source/api_reference/forecasting.rst`

File: sktime/forecasting/compose/_column_ensemble.py (hunk 11 above)
Comment (GuzalBulatova) [reply sktime/forecasting/compose/_column_ensemble.py:148]: Yes, exactly

File: sktime/forecasting/compose/_column_ensemble.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+            The forecasters horizon with the steps ahead to to predict.
+        X : pd.DataFrame, optional (default=None)
+            Exogenous variables are ignored.
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+        names, forecasters, indices = self._check_forecasters()
+        if len(set(indices)) != len(indices):
```
Comment (mloning) [inline sktime/forecasting/compose/_column_ensemble.py:91]: One last comment, perhaps the index checks can go into the `_check_forecasters` method?

**Comment by GuzalBulatova** [comment]
Thanks for your review, @fkiraly! I'm setting `requires-fh-in-fit` to be `True` and changing author strings. 

As for `_update` - in another PR, regarding the `EnsembleForecaster` [#1190](https://github.com/alan-turing-institute/sktime/pull/1190/files) I've moved the method to the parent class, `_HeterogenousEnsembleForecaster`. So once 1190 is merged both `Ensemble` and `ColumnEnsemble` should inherit it from there.

**Review by mloning — COMMENTED** [review]
This looks good to me.

My only question is about the tags. Technically, it should be `"requires-fh-in-fit": False`, no? Because the ensemble class itself does not do anything with the forecasting horizon in fit. Only if one of the ensemble members (`forecasters`) needs it in `fit`, does the ensemble object also require it in `fit`. 

@fkiraly do you have some example code for how the updating would work?

**Review by mloning — APPROVED** [review]
I think all changes have been addressed - again great work @GuzalBulatova! I'll merge this now. 🎉

**Pull request merged**